src/interface/types.rs

```diff
@@ -251,20 +297,174 @@ pub fn get_pipearray<'a>(union_argument: Arg) -> Result<&'a mut PipeArray, i32>
+pub fn get_sockpair<'a>(union_argument: Arg) -> Result<&'a mut SockPair, i32> {
+    let pointer = unsafe{union_argument.dispatch_sockpair};
+    if !pointer.is_null() {
+        return Ok(unsafe{&mut *pointer});
+    }
+    return Err(syscall_error(Errno::EFAULT, "dispatcher", "input data not valid"));
 }
 
-//EPOLL
-#[derive(Copy, Clone, Debug)]
-#[repr(C)]
-pub struct EpollEvent {
-    pub events: u32,
-    pub fd: i32 
-    //in native this is a union which could be one of a number of things
-    //however, we only support EPOLL_CTL subcommands which take the fd
+pub fn fd_set_to_hashset(union_argument: Arg, nfds: i32) -> Result<interface::RustHashSet<i32>, i32> {
+    let pointer = unsafe{union_argument.dispatch_mutcbuf};
+    let mut hashset = interface::RustHashSet::new();
+    if !pointer.is_null() {    
+        for i in 0..nfds {
+            let byte_offset = i / 8;
+            let bit_offset = i & 0b111;
+            let byte_ptr = pointer.wrapping_offset(byte_offset as isize);
+
+            //check whether the bit_offsetth bit of the byte_offsetth byte in the fd_set is set
+            if (unsafe{*byte_ptr}) & (1 << bit_offset) != 0 {
+                //if so, add it to our hashset
+                hashset.insert(i);
+            }
+        }
+    }
+    return Ok(hashset);
+}
+pub fn copy_out_to_fd_set(union_argument: Arg, nfds: i32, hashset: interface::RustHashSet<i32>) {
+    let pointer = unsafe{union_argument.dispatch_mutcbuf};
+    if pointer.is_null() {return;} //do nothing if it's null
+    for i in 0..nfds {
+        let byte_offset = i / 8;
+        let bit_offset = i & 0b111;
+        let byte_ptr = pointer.wrapping_offset(byte_offset as isize);
+
+        if hashset.contains(&i) {
+            //if it's in the hash set, set the bit ot 1
+            unsafe{*byte_ptr |= 1 << bit_offset;}
+        } else {
+            //else, set the bit to 0
+            unsafe{*byte_ptr &= !1 << bit_offset;}
+        }
+    }
+}
+
+pub fn get_sockaddr(union_argument: Arg, addrlen: u32) -> Result<interface::GenSockaddr, i32> {
+    let pointer = unsafe{union_argument.dispatch_constsockaddrstruct};
+    if !pointer.is_null() {    
+        let tmpsock = unsafe{&*pointer};
+        match tmpsock.sa_family {
+            /*AF_INET*/ 2 => {
+                if addrlen < size_of::<interface::SockaddrV4>() as u32 {
+                    return Err(syscall_error(Errno::EINVAL, "dispatcher", "input length too small for family of sockaddr"));
+                }
+                let v4_ptr = pointer as *mut interface::SockaddrV4;
+                return Ok(interface::GenSockaddr::V4(unsafe{*v4_ptr}));
+            }
+            /*AF_INET6*/ 30 => {
+                if addrlen < size_of::<interface::SockaddrV6>() as u32 {
+                    return Err(syscall_error(Errno::EINVAL, "dispatcher", "input length too small for family of sockaddr"));
+                }
+                let v6_ptr = pointer as *mut interface::SockaddrV6;
```

Comment:
At the beginning, we use the Union argument's const struct for the addresses. Then, we export them as a dereferenced mutable SockAddrV6. Should the pointer start as a dispatch_constsockaddrstruct? For instance, looking at the bind system call, it seems to use this version of the Union. But, in the tests, we input a mutable socketaddress. I might be wrong, but this feels off. Slack me if that made no sense.

Reply:
Ok right. Only in recvfrom do we actually need to make this sockaddr mutable. It is entirely ok to borrow a mutable sockaddr as const, but it is not as ok to then borrow the dereference of a const pointer as mutable like we do in recvfrom. We would ideally make 2 functions: one that makes it const which we'd use in all other functions and one that uses a mutsockaddr that we use only in recvfrom. Does that answer your question/issue? Is there anything more?